### ISSUE TITLE
Access beta map of FirstLevelModel

### ISSUE THREAD

**Issue by mathias-sm** [nilearn/nilearn#2544]
I can't find how to access the coefficients of a fitted FirstLevelModel at a given voxel. For early exploration of a dataset, having a method to access them would be useful (much like there's already `predicted`, `r_squared`, etc.). Starting from this:

```python
import pandas as pd
from nilearn.datasets import fetch_spm_auditory
from nilearn.image import concat_imgs
from nilearn.glm.first_level import FirstLevelModel

subject_data = fetch_spm_auditory()
fmri_img = concat_imgs(subject_data.func)
events = pd.read_table(subject_data['events'])
fmri_glm = FirstLevelModel(t_r=7, minimize_memory=False).fit(fmri_img, events)
```

`fmri_glm.design_matrices_[0].shape` tells me I have 16 regressors, how would one go at looking at the betas say at coordinates `(x, y, z) = (20, 20, 20)`? I tried the following but (i) it's not really straightforward, and (ii) it looks like some information is lost when stored in the `results_` dict?

```python
key = fmri_glm.masker_.inverse_transform(fmri_glm.labels_[0]).get_fdata()[x, y, z]
fmri_glm.results_[0][key].theta
```

Now, ```fmri_glm.results_[0][key].theta.shape``` gives me (16, 2204), the 16 looks like the betas I'm looking for but I don't know which of the 2204 vector I'm looking for.

**Comment by dangom** [comment]
Try:
```python
fmri_glm.compute_contrast("regressor_as_named_in_the_design_matrix", output_type="effect_size")
```

**Comment by bthirion** [comment]
thx @dangom !

**Comment by mathias-sm** [comment]
Thanks. So if I want to access the `effect_size` of all regressors at a given voxel, I should loop over the regressors, do this for each and extract the desired value — and if I want to look at various voxels often then merge the results in a `Nifti1Image` object to be saved?

Following the code of `compute_contrast` I ended up looking at the following snippet here::
https://github.com/nilearn/nilearn/blob/cfcb4fffbcf79a72160c5ad96348b782cbb533f6/nilearn/glm/contrasts.py#L71-L74

I came up with the following solution (assuming a single run), but I'm not quite sure that I'm doing the right thing with the masking:
```python
effect = np.zeros((len(fmri_glm.design_matrices_[0].columns), fmri_glm.labels_[0].size))
for label_ in fmri_glm.results_[0]:
    label_mask = fmri_glm.labels_[0] == label_
    resl = fmri_glm.results_[0][label_].theta
    effect[:, label_mask] = resl
```

From this indeed `fmri_glm.masker_.inverse_transform(effect).shape` returns the (expected) `(x, y, z, n_predictor)`

**Comment by bthirion** [comment]
This is correct indeed.

**Comment by mathias-sm** [comment]
Thanks for the quick help

**Comment by dangom** [comment]
Following up on this thread, I reckon this should be a common enough operation to warrant it's own simple interface. Would it not make sense to have a FirstLevelModel.get_effect_sizes() method?

**Comment by bthirion** [comment]
I know that people like to access these parameters, yet accessing them without knowing the corresponding variance/uncertainty/variance is futile and misleading. In particular because the raw parameter values depend on the arbitrary scalings of the design matrix and of the time series. Personally, I think that studentized values are less misleading.

**Comment by psychNerdJae** [comment]
Sorry to bump an old issue, but I wanted to give some user feedback on this – I actually have the URL to this GH issue copy/pasted in all of the scripts where I need to access beta maps, as a naive search for ["beta map"](https://nilearn.github.io/stable/search.html?q=beta+map&check_keywords=yes&area=default) on the website doesn't seem to return relevant documentation, and I'm otherwise constantly forgetting how to do it. The documentation returns more relevant information when the user knows to look for ["contrast effect size"](https://nilearn.github.io/stable/search.html?q=contrast+effect+size&check_keywords=yes&area=default#), but I think it is probably a less intuitive search term for most users, and as far as I'm aware, only [one example article](https://nilearn.github.io/stable/auto_examples/00_tutorials/plot_single_subject_single_run.html?highlight=contrast+effect+size#detecting-voxels-with-significant-effects) explicitly links the contrast effect size with the word "beta."

For me, a common use case is when I'm doing RSA, where I want to take the raw parameter estimates corresponding to a set of conditions/regressors, and to compute the pattern dissimilarity between them. While some variants of RSA can be performed using univariate noise normalization (i.e., studentized estimates), other variants rely on the raw betas because they use multivariate noise normalization (i.e. [Mahalanobis distances](https://rsatoolbox.readthedocs.io/en/latest/distances.html#mahalanobis-distance) where the noise covariance is estimated from the residuals of the first-level analysis).

While I agree that it can be misleading in many cases to access beta maps without taking variance into consideration, I also think updated documentation could help users better understand how to access beta maps, and when it is (in)appropriate to do so.

Also, as a side note, thanks to all of the nilearn developers – this is a great set of tools, which I enjoy using. :)

**Comment by bthirion** [comment]
Sorry, my previous answer was possibly a bit abrupt. Would you open a PR and implement the method ?

**Comment by alexisthual** [comment]
Hey!
Sorry for bumping this old issue as well. I often get questions about how to get the beta values of a fitted regression. Maybe it would be worth reopening this issue! @psychNerdJae do you think you'd be keen on opening a PR to solve this? :blush:

**Comment by psychNerdJae** [comment]
Hi @bthirion @alexisthual, thanks for the feedback, and sorry for the delayed response! Unfortunately, I'm not quite confident enough in my development skills right this moment to open a PR/implement the method. I'm always happy to learn, if someone from the team is willing to point me to some beginner-friendly resources where I could learn how to do this – but I'll be a bit swamped trying to wrap up some dissertation projects for some time.

**Comment by alexisthual** [comment]
I understand it can seem overwhelming, but we're here to help! I'd be happy to help review your PR if you open it :slightly_smiling_face: Maybe someone else could point you to some useful resources but I can't really think of any right now. I think you mostly need to jump in with both feet haha

**Comment by bthirion** [comment]
Reading the thread again:
* I'm not against giving access to beta maps, but to be frank, beta maps without noise estimates are useless if not misleading. For instance, for RSA estimates, I would really recommend using studentized versions
* Regarding terminology, it is important that nilearn code remains understandable by people unfamiliar with neuroimaging jargon. Beta has lots of different meanings depending on the field you come from. But I agree that we should clarify this in the documentation
* @psychNerdJae I'd be super happy if you can contribute this indeed...

**Comment by alexisthual** [comment]
Yes, I'm still confused as to what people mean when they say "betas". Some refer to the exact regression coefficients, others mean the normalized values of these coefficients.

Maybe instead of making a wrapper function to make them accessible, we should clarify the documentation and write the explicit formulas of what we are computing, both in the method docstrings and in the user guide (in particular, I'm surprised that we didn't include any mathematical formulation of what we're computing in the [GLM section of the user guide](https://nilearn.github.io/stable/glm/first_level_model.html)).

More specifically, I see at least two points where there's room for improvement:
* `glm.compute_contrast()` and `glm.first_level.FirstLevelModel.compute_contrast()` do different things, which I think is super mis-leading. If I understand correctly, the second function is actually more generic than the first one, and may deserve a different name?
* valid values for `output_type` in `glm.first_level.FirstLevelModel.compute_contrast()` are hard to grasp: `z_scores` and `p_value` are understandable, but what are `effect_size` and `stat` for instance? I think we should very explicitly write what is computed.

Finally, maybe we could modify an existing example to include a few lines explaining how to fetch the regression coefficients, why they should be normalized, and how to normalize them.

**Comment by psychNerdJae** [comment]
For whatever one user's input is worth, I agree with @alexisthual's ideas for improvement – this is closer to what I had in mind when I wrote my first post in this thread.

I should clarify that I completely agree with @bthirion that "classic" approaches to RSA (computing a dissimilarity matrix based on correlation distance) should be done with t-maps, not beta-maps. There is real danger in using the raw estimates (to me, "beta" refers to the exact regression estimates immediately after fitting a first-level GLM) rather than some form of noise-normalized estimate.

However, for researchers interested in using multivariate noise normalization methods like Mahalanobis distances ("multivariate" as opposed to the univariate noise normalization provided by studentized t-maps), and particularly crossvalidated Mahalanobis distances (i.e. the "crossnobis" estimator), it's essential that the "signal" (beta estimates) is on the same scale / uses the same units as the "noise" (residuals). It's easy enough to do this using existing nilearn functions, except that the terminology is sufficiently ambiguous that users can't be sure they're getting the estimates they need without referencing the source code.

I don't have a good sense for how niche this use-case is (or how many nilearn users are explicitly interested in RSA), but just wanted to provide my perspective. Thanks again for the thoughtful engagement!

**Comment by bthirion** [comment]
@alexisthual I agree with your second point, I think I don't with the first: To me the two functions do the same thing ?
@psychNerdJae I definitely agree that beta-maps should be easily accessible. Simply beware that they do not immediately compare to residuals, because the design matrix scaling is arbitrary.

**Comment by alexisthual** [comment]
> To me the two functions do the same thing ?

I might be lost haha, but they don't have the same API so it feels to me as if they were computing different things. For instance, I don't understand if `glm.compute_contrast()` can return a `stat` output (I guess it's a "studentized" beta) as `glm.FirstLevelModel.compute_contrast()` would.
My humble opinion is that we should try to unify these two APIs.

**Comment by bthirion** [comment]
Agreed, they should yield the same outputs.

**Comment by psychNerdJae** [comment]
A hopefully quick clarification question for @bthirion: I'm not quite sure what you mean by "[beta maps] do not immediately compare to residuals" – I took a look through the documentation ([this](https://nilearn.github.io/stable/auto_examples/04_glm_first_level/plot_predictions_residuals.html?highlight=residuals) seems to be the most relevant result?) and the source code for clues, but came up short. [It seems](https://github.com/nilearn/nilearn/blob/9c29e3376/nilearn/glm/first_level/first_level.py#L605:L610) like all data scaling occurs immediately before fitting the GLM, regardless of what scaling method is used – in which case, wouldn't this imply that the residuals would be on the same scale as the regression coefficients?

If it happens to be at all relevant to the conversation, I'm using "grand-mean" signal-scaling (along both time and voxels) to emulate SPM's first-level fitting procedure – I mention this only because [the most relevant example](https://nilearn.github.io/stable/auto_examples/04_glm_first_level/plot_predictions_residuals.html?highlight=residuals) unexpectedly sets `signal_scaling=False` despite there being [good reasons](https://neurostars.org/t/grand-mean-scaling-still-unclear-about-why/4796) to scale the data. Not sure if that was an idiosyncratic choice for that particular example, or if this is related to the comment about the beta/residual scales being incomparable?

Also happy to start a new issue if this is getting off-topic, though I think the question about residual scaling follows directly from questions about how beta-maps are mathematically defined / returned.

**Comment by bthirion** [comment]
The scaling that you allude to is the signal scaling, which is done prior to signal fitting.

But the design matrix scaling comes into play, when you want to compare beta magnitude to residual manitude, since  beta coefficients are inversely proportional to the design matrix.
Nilearn convention is that a beta of 1 corresponds to a 1 signal unit (so typically, 1% of the baseline of the value) increase in response to the stimulation. But this only applies to regressors that are obtained from events descriptors by convolution with an hrf. Nothing prevents users from providing other regressors  with arbitrary scaling or adopting a different scaling if they prefer.
 Hope that this is clearer now...

**Comment by psychNerdJae** [comment]
Yes, that makes perfect sense – thank you!

**Comment by ymzayek** [comment]
Thanks all for the discussion on this. I think we can definitely start by improving the documentation on this as stated in https://github.com/nilearn/nilearn/issues/2544#issuecomment-1426787383 as it seems like there is a consensus on that. We can consider adding [this simple example of accessing a beta map](https://github.com/nilearn/nilearn/issues/2544#issuecomment-663197447) first along with an explanation of when it is bad practice and why. And following that we can rethink if providing a method to simplify this for all regressors is still necessary? 
@psychNerdJae if you want to work on this I think the easiest way to start is to tackle the documentation. We can also split it further to smaller tasks if that helps. You can take a look at our [contributing docs](https://nilearn.github.io/stable/development.html#how-to-contribute-to-nilearn) to help you get started with the PR workflow.
